src/libstore/store-api.cc

```diff
@@ -1319,6 +1320,21 @@ std::shared_ptr<Store> openFromNonUri(const std::string & uri, const Store::Para
             return std::make_shared<LocalStore>(params);
         else if (pathExists(settings.nixDaemonSocketFile))
             return std::make_shared<UDSRemoteStore>(params);
+        #if __linux__
+        else if (!pathExists(stateDir) && params.empty() && getuid() != 0) {
```

Comment:
Could this instead try to create the state dir and only fallback to a chroot store if it can't? That would avoid issues like https://github.com/NixOS/nix/issues/6732